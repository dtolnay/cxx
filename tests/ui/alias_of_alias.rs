// Aliases of aliases don't work if the namespace of any intermediate alias
// differs from the original bridge's namespace. This is because the ExternType
// trait being checked against will use the original namespace, which the final
// alias does not know. However, if not for the ExternType check, this would
// work because of the C++ "using" aliases generated from each bridge.
//
// This can be made to work if needed - either through an attribute on the alias
// as a "manual" fix, or potentially through some creativity with ExternType
// (like defining a related type that implements ExternType and checking against
// that instead of the actual type) - but seems like a rare case.

#[cxx::bridge(namespace = here)]
mod here {
    extern "C" {
        type C;
    }
}

// Rustfmt mangles the extern type alias.
// https://github.com/rust-lang/rustfmt/issues/4159
#[rustfmt::skip]
#[cxx::bridge(namespace = there)]
#[cxx::alias_namespace(crate::here = here)]
mod there {
    extern "C" {
        type C = crate::here::C;
    }
}

// Rustfmt mangles the extern type alias.
// https://github.com/rust-lang/rustfmt/issues/4159
#[rustfmt::skip]
#[cxx::bridge(namespace = anywhere)]
#[cxx::alias_namespace(crate::there = there)]
mod anywhere {
    extern "C" {
        type C = crate::there::C;
    }
}

fn main() {}
