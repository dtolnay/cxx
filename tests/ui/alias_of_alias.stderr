error[E0271]: type mismatch resolving `<here::C as ExternType>::Id == (t, h, e, r, e, (), cxx::C)`
   --> $DIR/alias_of_alias.rs:33:1
    |
33  | #[cxx::bridge(namespace = anywhere)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected a tuple with 7 elements, found one with 6 elements
    |
   ::: $WORKSPACE/src/extern_type.rs
    |
    | pub fn verify_extern_type<T: ExternType<Id = Id>, Id>() {}
    |                                         ------- required by this bound in `verify_extern_type`
    |
    = note: expected tuple `(t, h, e, r, e, (), cxx::C)`
               found tuple `(h, e, r, e, (), cxx::C)`
